File: libs/shared-types/python-src/debrief/validators/track_validator.py
"""
Manual validators for Track features
These validators work with the generated types and provide additional validation logic
"""

from typing import Any, Dict, List, Union
from datetime import datetime

try:
    from ..types.track import TrackFeature
except ImportError:
    # If import fails, define a minimal type for validation
    class TrackFeature:  # type: ignore
        pass


def validate_timestamps_length(feature: Dict[str, Any]) -> bool:
    """
    Validates that timestamps array length matches coordinate points count
    This is the critical cross-field validation not covered by JSON Schema
    """
    properties = feature.get("properties", {})
    timestamps = properties.get("timestamps")

    if not timestamps:
        return True  # timestamps are optional

    geometry = feature.get("geometry", {})
    coordinates = geometry.get("coordinates", [])
    geometry_type = geometry.get("type")

    # Handle LineString geometry
    if geometry_type == "LineString":
        return len(timestamps) == len(coordinates)

    # Handle MultiLineString geometry
    if geometry_type == "MultiLineString":
        # Calculate total points across all LineStrings
        total_points = sum(len(line_string) for line_string in coordinates)
        return len(timestamps) == total_points

    return False


def validate_track_feature(feature: Any) -> bool:
    """
    Validates that track feature has required properties and valid structure
    """
    if not isinstance(feature, dict):
        return False

    # Check basic GeoJSON structure
    if feature.get("type") != "Feature":
        return False

    feature_id = feature.get("id")
    if feature_id is None:
        return False  # id is required

    # Check geometry
    geometry = feature.get("geometry")
    if not isinstance(geometry, dict):
        return False

    valid_geometry_types = ["LineString", "MultiLineString"]
    if geometry.get("type") not in valid_geometry_types:
        return False

    coordinates = geometry.get("coordinates")
    if not isinstance(coordinates, list):
        return False

    # Check properties
    properties = feature.get("properties")
    if not isinstance(properties, dict):
        return False

    # Validate timestamps if present
    timestamps = properties.get("timestamps")
    if timestamps is not None:
        if not isinstance(timestamps, list):
            return False

        # Check each timestamp is valid
        for timestamp in timestamps:
            if isinstance(timestamp, str):
                try:
                    datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
                except ValueError:
                    return False
            elif not isinstance(timestamp, datetime):
                return False

        # Apply timestamps length validation
        if not validate_timestamps_length(feature):
            return False

    return True


def validate_linestring_coordinates(coordinates: Any) -> bool:
    """
    Validates coordinate structure for LineString
    """
    if not isinstance(coordinates, list):
        return False

    if len(coordinates) < 2:
        return False  # LineString must have at least 2 points

    for coord in coordinates:
        if not isinstance(coord, list):
            return False
        if len(coord) < 2 or len(coord) > 3:
            return False
        if not all(isinstance(val, (int, float)) for val in coord):
            return False

    return True


def validate_multilinestring_coordinates(coordinates: Any) -> bool:
    """
    Validates coordinate structure for MultiLineString
    """
    if not isinstance(coordinates, list):
        return False

    if len(coordinates) < 1:
        return False  # MultiLineString must have at least 1 LineString

    return all(validate_linestring_coordinates(line_string) for line_string in coordinates)


def validate_track_feature_comprehensive(feature: Any) -> Dict[str, Union[bool, List[str]]]:
    """
    Comprehensive track feature validation
    """
    errors = []

    # Basic structure checks (without timestamp validation)
    if not isinstance(feature, dict):
        errors.append("Feature must be a dictionary")
        return {"is_valid": False, "errors": errors}

    if feature.get("type") != "Feature":
        errors.append('Feature type must be "Feature"')

    if feature.get("id") is None:
        errors.append("Feature must have an id")

    geometry = feature.get("geometry")
    if not isinstance(geometry, dict):
        errors.append("Feature must have a geometry object")
    else:
        geometry_type = geometry.get("type")
        if geometry_type not in ["LineString", "MultiLineString"]:
            errors.append("Track geometry must be LineString or MultiLineString")

        coordinates = geometry.get("coordinates", [])
        if not isinstance(coordinates, list):
            errors.append("Geometry must have coordinates array")
        else:
            # Detailed coordinate validation
            if geometry_type == "LineString":
                if not validate_linestring_coordinates(coordinates):
                    errors.append("Invalid LineString coordinates")
            elif geometry_type == "MultiLineString":
                if not validate_multilinestring_coordinates(coordinates):
                    errors.append("Invalid MultiLineString coordinates")

    properties = feature.get("properties")
    if not isinstance(properties, dict):
        errors.append("Feature must have a properties object")

    # Timestamps validation - check this regardless of other errors
    if not validate_timestamps_length(feature):
        errors.append("Timestamps array length does not match coordinate points count")

    return {"is_valid": len(errors) == 0, "errors": errors}


def is_valid_datetime_string(date_str: str) -> bool:
    """
    Validates that string is a valid ISO datetime
    """
    try:
        datetime.fromisoformat(date_str.replace("Z", "+00:00"))
        return True
    except (ValueError, TypeError):
        return False

File: libs/shared-types/src/validators/track-validator.ts
/**
 * Manual validators for Track features
 * These validators work with the generated types and provide additional validation logic
 */

import type { DebriefTrackFeature } from '../index';

/**
 * Type-safe helper to check if a value is a non-null object
 */
function isObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object';
}

/**
 * Type-safe helper to access object properties
 */
function getObjectProperty(obj: Record<string, unknown>, key: string): unknown {
  return obj[key];
}

/**
 * Validates that timestamps array length matches coordinate points count
 * This is the critical cross-field validation not covered by JSON Schema
 */
export function validateTimestampsLength(feature: DebriefTrackFeature): boolean {
  if (!feature.properties || !feature.properties.timestamps) {
    return true; // timestamps are optional
  }

  if (!feature.geometry) return false;

  // Handle LineString geometry
  if (feature.geometry.type === "LineString") {
    const coordinates = (feature.geometry as any).coordinates;
    return feature.properties.timestamps.length === coordinates.length;
  }

  // Handle MultiLineString geometry
  if (feature.geometry.type === "MultiLineString") {
    const coordinates = (feature.geometry as any).coordinates;
    // Calculate total points across all LineStrings
    const totalPoints = (coordinates as Array<Array<number[]>>)
      .reduce((sum, lineString) => sum + lineString.length, 0);
    return feature.properties.timestamps.length === totalPoints;
  }
  
  return false;
}

/**
 * Validates that track feature has required properties and valid structure
 */
export function validateTrackFeature(feature: unknown): feature is DebriefTrackFeature {
  if (!isObject(feature)) {
    return false;
  }
  
  // Check basic GeoJSON structure
  if (getObjectProperty(feature, 'type') !== 'Feature') {
    return false;
  }
  
  const id = getObjectProperty(feature, 'id');
  if (id === null || id === undefined) {
    return false; // id is required (can be 0 but not null/undefined)
  }
  
  // Check geometry
  const geometry = getObjectProperty(feature, 'geometry');
  if (!isObject(geometry)) {
    return false;
  }
  
  const validGeometryTypes = ['LineString', 'MultiLineString'];
  const geometryType = getObjectProperty(geometry, 'type');
  if (typeof geometryType !== 'string' || validGeometryTypes.indexOf(geometryType) === -1) {
    return false;
  }
  
  const coordinates = getObjectProperty(geometry, 'coordinates');
  if (!Array.isArray(coordinates)) {
    return false;
  }
  
  // Check properties
  const properties = getObjectProperty(feature, 'properties');
  if (!isObject(properties)) {
    return false;
  }
  
  // Check required dataType discriminator
  if (getObjectProperty(properties, 'dataType') !== 'track') {
    return false;
  }
  
  // Validate timestamps if present
  const timestamps = getObjectProperty(properties, 'timestamps');
  if (timestamps !== undefined) {
    if (!Array.isArray(timestamps)) {
      return false;
    }
    
    // Check each timestamp is a valid date string or Date object
    for (const timestamp of timestamps) {
      if (typeof timestamp === 'string') {
        if (isNaN(Date.parse(timestamp))) {
          return false;
        }
      } else if (!(timestamp instanceof Date)) {
        return false;
      }
    }
    
    // Apply timestamps length validation
    if (!validateTimestampsLength(feature as unknown as DebriefTrackFeature)) {
      return false;
    }
  }
  
  return true;
}

/**
 * Validates coordinate structure for LineString
 */
export function validateLineStringCoordinates(coordinates: unknown): boolean {
  if (!Array.isArray(coordinates)) {
    return false;
  }
  
  if (coordinates.length < 2) {
    return false; // LineString must have at least 2 points
  }
  
  return coordinates.every(coord => 
    Array.isArray(coord) && 
    coord.length >= 2 && 
    coord.length <= 3 &&
    coord.every(val => typeof val === 'number')
  );
}

/**
 * Validates coordinate structure for MultiLineString
 */
export function validateMultiLineStringCoordinates(coordinates: unknown): boolean {
  if (!Array.isArray(coordinates)) {
    return false;
  }
  
  if (coordinates.length < 1) {
    return false; // MultiLineString must have at least 1 LineString
  }
  
  return coordinates.every(lineString => validateLineStringCoordinates(lineString));
}

/**
 * Comprehensive track feature validation
 */
export function validateTrackFeatureComprehensive(feature: unknown): {
  isValid: boolean;
  errors: string[];
} {
  const errors: string[] = [];
  
  // Basic validation
  if (!validateTrackFeature(feature)) {
    errors.push('Invalid track feature structure');
    return { isValid: false, errors };
  }
  
  const validatedFeature = feature as DebriefTrackFeature;
  
  // Detailed coordinate validation
  if (!validatedFeature.geometry) {
    errors.push('Geometry is required');
  } else if (validatedFeature.geometry.type === 'LineString') {
    if (!validateLineStringCoordinates(validatedFeature.geometry.coordinates)) {
      errors.push('Invalid LineString coordinates');
    }
  } else if (validatedFeature.geometry.type === 'MultiLineString') {
    if (!validateMultiLineStringCoordinates(validatedFeature.geometry.coordinates)) {
      errors.push('Invalid MultiLineString coordinates');
    }
  }
  
  // Timestamps validation
  if (!validateTimestampsLength(validatedFeature)) {
    errors.push('Timestamps array length does not match coordinate points count');
  }
  
  return {
    isValid: errors.length === 0,
    errors
  };
}